src/breakpoints.jl

```diff
@@ -1,25 +1,99 @@
+    function extract_function_from_method(m::Method)
+        sig = Base.unwrap_unionall(m.sig)
+        ft0 = sig.parameters[1]
+        ft = Base.unwrap_unionall(ft0)
+        if ft <: Function && isa(ft, DataType) && isdefined(ft, :instance)
+            return ft.instance
+        elseif isa(ft, DataType) && ft.name === Type.body.name
+            f = ft.parameters[1]
+        else
+            return ft
+        end
     end
+
+    framecode.scope isa Method || return false
+    meth = framecode.scope
+    bp.f isa Method && return meth === bp.f
+    bp.f === extract_function_from_method(meth) || return false
+    bp.sig === nothing && return true
+    return bp.sig <: meth.sig
+end
+
+"""
+    breakpoint(f, [sig], [line], [condition])
+
+Add a breakpoint to `f` with the specified argument types `sig`.¨
+If `sig` is not given, the breakpoint will apply to all methods of `f`.
+If `f` is a method, the breakpoint will only apply to that method.
+Optionally specify an absolute line number `line` in the source file; the default
+is to break upon entry at the first line of the body.
+Without `condition`, the breakpoint will be triggered every time it is encountered;
+the second only if `condition` evaluates to `true`.
+`condition` should be written in terms of the arguments and local variables of `f`.
+
+# Example
+```julia
+function radius2(x, y)
+    return x^2 + y^2
+end
+
+breakpoint(radius2, Tuple{Int,Int}, :(y > x))
+```
+"""
+function breakpoint(f, sig=nothing, line::Integer=0, condition::Condition=nothing)
+    sig !== nothing && (sig = Base.to_tuple_type(sig))
+    bp = BreakpointSignature(f, sig, line, condition, Ref(true), BreakpointRef[])
+    add_to_existing_framecodes(bp)
+    idx = findfirst(bp2 -> same_location(bp, bp2), _breakpoints)
+    idx === nothing ? push!(_breakpoints, bp) : (_breakpoints[idx] = bp)
     return bp
 end
+breakpoint(f, sig, condition::Condition) = breakpoint(f, sig, 0, condition)
+breakpoint(f, line::Integer, condition::Condition=nothing) = breakpoint(f, nothing, line, condition)
+breakpoint(f, condition::Condition) = breakpoint(f, nothing, 0, condition)
+
+
+"""
+    breakpoint(filename, line, [condition])
+
+Set a breakpoint at the specified file and line number.
+"""
+function breakpoint(file::String, line::Integer=0, condition::Condition=nothing)
+    maybe_source_file = Base.find_source_file(file)
+    if maybe_source_file === nothing
+        file = abspath(file)
+    end
+    bp = BreakpointFileLocation(Symbol(file), line, condition, Ref(true), BreakpointRef[])
+    add_to_existing_framecodes(bp)
+    idx = findfirst(bp2 -> same_location(bp, bp2), _breakpoints)
+    idx === nothing ? push!(_breakpoints, bp) : (_breakpoints[idx] = bp)
+    return bp
+end
+
+function framecode_matches_breakpoint(framecode::FrameCode, bp::BreakpointFileLocation)
+    framecode.scope isa Method || return false
+    meth = framecode.scope
+    meth.file == bp.file || return false
```

Comment:
I wonder if we should be more relaxed here. For example, you might just want to bother writing a full path to it. Perhaps `endswith`? Then you can add a breakpoint to `foo.jl` if it is in `Package/src/foo.jl`?

Reply:
That is an attractive thought, it might also eliminate path weirdness (symlinks etc). OTOH I have been bit sometimes by Julia's `edit` if I happen to be in a directory that has a file of the same name as one of Julia's own base files. If lots of packages have the same file (e.g., `precompile.jl`) will this cause trouble?

Reply:
Maybe, but you always have the option to specify the full path if you find it matches places where you don't want it to.